src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs

```diff
@@ -0,0 +1,289 @@
+            _projectFileModelWatcher.InitializeModelWatcher();
+
+            // Finally, move to the editor open state
+            lock (_lock)
+            {
+                _currentState = EditorState.EditorOpen;
+            }
+        }
+
+        // Teardown Logic
+
+        /// <summary>
+        /// Tells the editor to call close on the existing window frame.
+        /// </summary>
+        /// <returns>True if the close was successful, false if the window is still open</returns>
+        public async Task<bool> CloseWindowAsync()
+        {
+            // We can't switch to editor closing here because there's still the chance that some other action can cancel
+            // the editor closing, without notifying us. A later callback will get hit once the close is guaranteed to happen.
+            await _threadingService.SwitchToUIThread();
+            return _windowFrame.CloseFrame((uint)__FRAMECLOSE.FRAMECLOSE_PromptSave) == VSConstants.S_OK;
+        }
+
+        /// <summary>
+        /// Starts the process of closing the project file editor, if one is open currently.
+        /// </summary>
+        public async Task DisposeEditorAsync()
+        {
+            lock (_lock)
+            {
+                if (_currentState == EditorState.NoEditor) return;
+
+                // Checking for potential dirty state and asking if the user wants to save their changes will have already occurred at this point.
+                // Just go to EditorClosing.
+                _currentState = EditorState.EditorClosing;
+            }
+
+            _projectFileModelWatcher?.Dispose();
+            _textBufferStateListener?.Dispose();
+            if (_frameEventsListener != null)
+            {
+                await _frameEventsListener.DisposeAsync().ConfigureAwait(false);
+            }
+            _textBufferManager?.Dispose();
+
+            _projectFileModelWatcher = null;
+            _frameEventsListener = null;
+            _textBufferStateListener = null;
+            _textBufferManager = null;
+
+            lock (_lock)
+            {
+                _currentState = EditorState.NoEditor;
+            }
+            return;
+        }
+
+        // Update Project File Logic
+
+        /// <summary>
+        /// Schedules the project file to be updated. This is called by the ProjectFileModelWatcher, which is within a Project write lock.
+        /// Because of the time-sensitive nature of that lock, instead of updating the file immediately we schedule an update as soon
+        /// as the JTF can run our code.
+        /// </summary>
+        public JoinableTask ScheduleProjectFileUpdate()
+        {
+            lock (_lock)
+            {
+                // If the current state is writing project file, we don't want to update now, as the project will not have been fully
+                // reloaded yet. We'll get called back again after the ProjectReloadManager is finished reloading the project.
```

Comment:
`UnconfiguredProjectAdvanced.ChangingProjectFile` will get triggered again because the project will be live reloaded.